### ISSUE TITLE
Compiler panic on `cx.tcx.fn_sig(fn_def_id)`

### ISSUE THREAD

**Issue by yoav-lavi** [rust-lang/rust-clippy#8654]
Ran into a compiler panic when running the following:

```rust
use clippy_utils::diagnostics::span_lint_and_sugg;
use if_chain::if_chain;
use rustc_ast::Attribute;
use rustc_errors::Applicability;
use rustc_hir::intravisit::FnKind;
use rustc_hir::{
    Arm, Block, Body, Expr, ExprKind, FieldDef, FnDecl, ForeignItem, GenericArg, GenericParam, Generics, HirId,
    ImplItem, InferArg, Item, Lifetime, Local, Mod, Param, Pat, Path, PolyTraitRef, Stmt, TraitBoundModifier,
    TraitItem, Ty, Variant, VariantData, WherePredicate
};
use rustc_lint::{LateContext, LateLintPass};
use rustc_session::{declare_lint_pass, declare_tool_lint};
use rustc_span::{Span, Symbol, def_id::LocalDefId};
use rustc_span::symbol::sym;

declare_clippy_lint! {
    /// ### What it does
    ///
    /// Detects cases of `&String::new()` being passed as an argument to a function expecting `&str`
    ///
    /// ### Why is this bad?
    ///
    /// `&String::new()` is longer and may be confusing
    ///
    /// ### Example
    /// ```rust
    /// "12345".split(&String::new());
    /// ```
    /// Use instead:
    /// ```rust
    //  "12345".split("");
    /// ```
    #[clippy::version = "1.61.0"]
    pub UNNECESSARY_STRING_NEW,
    style,
    "Detects cases of `&String::new()` being passed as an argument to a function expecting `&str`"
}
declare_lint_pass!(UnnecessaryStringNew => [UNNECESSARY_STRING_NEW]);

impl<'tcx> LateLintPass<'tcx> for UnnecessaryStringNew {
    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
        if_chain! {
            if let ExprKind::Call(callee, args) = expr.kind;
            if let ExprKind::Path(_) = callee.kind;
            let callee_ty = cx.typeck_results().expr_ty_adjusted(callee);
            let fn_def_id = cx.tcx.hir().local_def_id(callee.hir_id);
            then {
                println!("{:#?}", cx.tcx.fn_sig(fn_def_id));
            }
        }
    }
}
```

On this file (via `cargo dev lint tests/ui/unnecessary_string_new.rs`):

```rust
// run-rustfix

#![warn(clippy::unnecessary_string_new)]

fn dummy(_value: &str) {}

fn main() {
    // should be linted
    dummy(&String::new());

    // should not be linted
    dummy("");
}
```

Message: 

```
error: internal compiler error: /rustc/9f4dc0b4db892271cd0dada6e072775b5b5d6b1e/compiler/rustc_middle/src/hir/map/mod.rs:193:13: local_def_id: no entry for `HirId { owner: DefId(0:4 ~ unnecessary_string_new[eebc]::main), local_id: 2 }`, which has a map of `Some(Expr(Expr { hir_id: HirId { owner: DefId(0:4 ~ unnecessary_string_new[eebc]::main), local_id: 2 }, kind: Path(Resolved(None, Path { span: tests/ui/unnecessary_string_new.rs:9:5: 9:10 (#0), res: Def(Fn, DefId(0:3 ~ unnecessary_string_new[eebc]::dummy)), segments: [PathSegment { ident: dummy#0, hir_id: Some(HirId { owner: DefId(0:4 ~ unnecessary_string_new[eebc]::main), local_id: 1 }), res: Some(Err), args: None, infer_args: true }] })), span: tests/ui/unnecessary_string_new.rs:9:5: 9:10 (#0) }))`

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1261:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.1.61 (cf1e2e9c1 2022-04-06)

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error
```

`rustup toolchain list`:
```
stable-aarch64-apple-darwin (default)
nightly-2022-03-14-aarch64-apple-darwin
nightly-2022-03-24-aarch64-apple-darwin
nightly-aarch64-apple-darwin
1.59.0-aarch64-apple-darwin
```

Note that this is from running `clippy` with a new lint from the master branch via `cargo dev lint`

**Comment by Jarcho** [comment]
That doesn't get the `DefId` of the function. That would be 
```rust
if_chain! {
    if let ExprKind::Call(callee, args) = expr.kind;
    if let ExprKind::Path(ref path) = callee.kind;
    if let Some(def_id) = cx.qpath_res(path, callee.hir_id);
    then { /* use def_id */ }
}
```
There's also `clippy_utils::ty::expr_sig` which can get the signature of any callable object, rather than just functions.

---

Also this isn't really the appropriate place for this. For help with writing a lint you should either use zulip, or, since you have a PR open, ask there.

**Comment by yoav-lavi** [comment]
@Jarcho thanks for the response, I didn't mean this as a question but rather thought that this was a bug (since clippy uses compiler internals it makes sense that it could trigger a compiler panic, but I wasn't sure if what I was seeing was an actual issue or not)
